include/led.h

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _ZEPHYR_LED_H
+#define _ZEPHYR_LED_H
+
+/**
+ * @file
+ * @brief Public LED driver APIs
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+/**
+ * @brief Driver specific representation of a LED device
+ */
+
+struct led_data {
+	/* Minimum acceptable time period (in ms) supported by the device */
+	u32_t min_period;
+	/* Maximum acceptable time period (in ms) supported by the device */
+	u32_t max_period;
+	/* Minimum acceptable brightness (in percent) supported by the device */
+	u32_t min_brightness;
+	/* Maximum acceptable brightness (in percent) supported by the device */
+	u32_t max_brightness;
+};
+
+/**
+ * @typedef led_api_blink()
+ * @brief Callback API for blinking a LED
+ *
+ * @see led_blink() for argument descriptions.
+ */
+typedef int (*led_api_blink)(struct device *dev, u32_t led, u32_t period);
+
+/**
+ * @typedef led_api_blink_oneshot()
+ * @brief Callback API for oneshot blink of a LED
+ *
+ * @see led_blink_oneshot() for argument descriptions.
+ */
+typedef int (*led_api_blink_oneshot)(struct device *dev, u32_t led);
+
+/**
+ * @typedef led_api_set_brightness()
+ * @brief Callback API for setting brightness of a LED
+ *
+ * @see led_set_brightness() for argument descriptions.
+ */
+typedef int (*led_api_set_brightness)(struct device *dev, u32_t led,
+				      u32_t value);
+/**
+ * @typedef led_api_on()
+ * @brief Callback API for turning on an LED
+ *
+ * @see led_on() for argument descriptions.
+ */
+typedef int (*led_api_on)(struct device *dev, u32_t led);
+
+/**
+ * @typedef led_api_off()
+ * @brief Callback API for turning off an LED
+ *
+ * @see led_off() for argument descriptions.
+ */
+typedef int (*led_api_off)(struct device *dev, u32_t led);
+
+/**
+ * @brief LED driver API
+ *
+ * This is the mandatory API any LED driver needs to expose.
+ */
+struct led_driver_api {
+	led_api_blink blink;
+	led_api_blink_oneshot blink_oneshot;
+	led_api_set_brightness set_brightness;
+	led_api_on on;
+	led_api_off off;
+};
+
+/**
+ * @brief Blink an LED
+ *
+ * This routine blinks an LED for the given time period continuously
+ *
+ * @param dev LED device
+ * @param led LED channel/pin
+ * @param period Period (in micro second) to blink an LED
```

Comment:
Two issues here.

1. I still don't understand the way "period" is being used in this API.
2. I don't think "period" makes sense to use in a general API; that seems to assume PWM.

By making definitions in terms of period, this seems to assume PWM is the underlying implementation. But this is a general purpose LED API. It should be possible to implement a driver using a true DAC, which doesn't have a period: you just set brightness directly. Or to implement a driver using a GPIO, which treats any brightness other than "off" into "full on" (following the "do your best" comment in the android HAL API I linked to in my first review).

Brightness is the LED phenomenon that people perceive which is being supported in this initial version. And there's already a PWM API, so adding an LED API that assumes PWM doesn't make sense to me.

I therefore think that definitions should be done in terms of brightness levels (since the API doesn't contemplate color channels yet).

Can you try thinking through what an implementation would look like for a few different types of driver (GPIO, DAC, PWM, digitally addressable "smart" LED controlled by SPI or I2C, etc.) and see what they have in common?

Reply:
Most of the LED hw are capable of blinking at a specific rate, that's why I used `period` notation to specify the blinking rate per second. If this creates ambiguity with PWM, we can switch to `delay on` / `delay_off`, what do you think?

Also, I need one more clarification from you. When you say `brightness` I assume it to be the brightness level (duty cycle in PWM terms) and what about the time period? Shouldn't we add that functionality too?

Reply:
> Most of the LED hw are capable of blinking at a specific rate, that's why I used period notation to specify the blinking rate per second.

Ah, OK, so `min_period` and `max_period` were about blinking rate per second, and are related to this routine? I didn't get that; IMO it isn't clear from the docstrings ("Minimum acceptable time period (in ms) supported by the device" doesn't mention blinking).

> If this creates ambiguity with PWM, we can switch to delay on / delay_off, what do you think?

My confusion was with the period of time over which a PWM wave oscillates through a single fundamental period: the amount of time the pin is HIGH plus the time it is LOW. I was assuming this was about PWMing rapidly to support dimming, rather than blinking at perceptible rates. Make sense?

> 
> Also, I need one more clarification from you. When you say brightness I assume it to be the brightness level (duty cycle in PWM terms) and what about the time period? Shouldn't we add that functionality too? And that's what this API is intended to do.

Now I'm confused about what you're trying to do again.

Setting "brightness" with a PWM-backed LED is something you can only do if the period is small enough (say a few milliseconds) so that people see the light get brighter and darker when you adjust the duty cycle higher and lower.

But if the period is long enough (say a few seconds) so that you can perceive the light turning on and off, adjusting the duty cycle has nothing to do with brightness in people's perception.

Looks like you're not trying to support brightness after all, just blinking at rates humans can perceive?

Reply:
I think there is a bit of confusion here. Let me make it clear of what I'm trying to achieve with this API.

So, I want to expose functions to set both brightness and blinking of the LED as below:

1. Brightness

Most of the LED devices are capable of setting brightness at the hardware level. Here brightness refers to the level an LED is glowing. The confusion came when I used PWM notations. So, just ignore duty cycle, frequency etc... and in pure LED terms `led_set_brightness` refers to setting the brightness level at which the LED can glow constantly.

2. Blinking

This refers to turning the LED on and off at a particular rate. This doesn't have anything to do with the level of brightness an LED is glowing. So, `led_blink` will just turn on and off the LED at the specified amount of time. Here, time at which LED turns on/off is passed in `period` parameter.

In a general usecase, you can set the brightness level of an LED to say 50 (again this value is specific to the hardware) and at this brightness you can turn it on and off once every 500ms. Otherwise, you can just set the LED to maximum brightness and turn it on and off every 500ms.

If you don't set the blinking time, it will glow constantly at the specified brightness value. And in future I have plans to implement software blinking so that a general purpose timer can be used to turn the LED on/off (here LED will glow at full brightness).